File: Color.py
# Bài 2: Coloring problem
V = 4  # Số lượng đỉnh trong đồ thị

def print_solution(color):
    print("Lời giải:")
    print(" ".join(map(str, color)))

def is_safe(v, graph, color, c):
    # Kiểm tra xem có thể tô màu c cho đỉnh v hay không
    for i in range(V):
        if graph[v][i] and c == color[i]:
            return False
    return True

def graph_coloring_util(graph, m, color, v):
    # Nếu đã tô màu xong cho tất cả các đỉnh
    if v == V:
        return True
    
    # Thử tô màu từ 1 đến m
    for c in range(1, m + 1):
        if is_safe(v, graph, color, c):
            color[v] = c
            
            # Đệ quy kiểm tra đỉnh kế tiếp
            if graph_coloring_util(graph, m, color, v + 1):
                return True
            
            # Quay lui nếu không tìm thấy lời giải
            color[v] = 0
    
    return False

def graph_coloring(graph, m):
    color = [0] * V  # Mảng lưu màu của các đỉnh
    
    if not graph_coloring_util(graph, m, color, 0):
        print("Noooo")
        return False
    
    print_solution(color)
    return True

if __name__ == "__main__":
    # Minh họa đồ thị
    # (3)---(2)
    #  |  /  |
    #  | /   |
    #  |/    |
    # (0)---(1)
    graph = [
        [0, 1, 1, 1],
        [1, 0, 1, 0],
        [1, 1, 0, 1],
        [1, 0, 1, 0],
    ]
    
    m = 3  # Số màu
    
    graph_coloring(graph, m)

File: Color.cpp
//Bài 2: Coloring problem
#include <iostream>
using namespace std;

//Số lượng đỉnh trong đồ thị
const int V = 4;

void printSolution(int color[]);

//Hàm kiểm tra điều kiện
bool isSafe(int v, bool graph[V][V], int color[], int c)
{
    for (int i = 0; i < V; i++)

    //Nếu có cạnh nối và màu trùng 
        if (graph[v][i] && c == color[i])
            return false;

    return true;
}

//Hàm tô màu
bool graphColoringUtil(bool graph[V][V], int m, int color[], int v)
{
    //Nếu đã tô màu xong cho tất cả các đỉnh
    if (v == V)
        return true;

    for (int c = 1; c <= m; c++)
    {//Thử màu 1 đến màu m
        //Gọi hàm kiểm tra => Nếu thỏa => Gán màu cho đỉnh v
        if (isSafe(v, graph, color, c))
        {
            //Gán màu
            color[v] = c;

            //Đê quỵ kiểm tra đỉnh kế tiếp
            if (graphColoringUtil(graph, m, color, v + 1))
                return true;
            
            //Quay lui nếu không tìm thấy lời giải
            color[v] = 0;
        }
    }
       return false;
}

bool graphColoring(bool graph[V][V], int m)
{
    int color[V]; //Mảng lưu màu của các đỉnh
    for (int i = 0; i < V; i++)
        color[i] = 0;

    //Gọi hàm tô màu bắt đầu từ đỉnh 0
    if (graphColoringUtil(graph, m, color, 0) == false)
    {
        //Nếu false => Noooo
        cout << "Noooo";
        return false;
    }

    //Gọi hàm in
    printSolution(color);
    return true;
}

//Hàm in lời giải
void printSolution(int color[])
{
    cout << "Loi giai: "<<endl;
    for (int i = 0; i < V; i++)
        cout << " " << color[i] << " ";

    cout << "\n";
}

int main()
{

    /* Minh họa đồ thị 
    (3)---(2)
    | / |
    | / |
    | / |
    (0)---(1)
    */
    bool graph[V][V] = {
        {0, 1, 1, 1},
        {1, 0, 1, 0},
        {1, 1, 0, 1},
        {1, 0, 1, 0},
    };

    //Số màu
    int m = 3;

    //Gọi hàm
    graphColoring(graph, m);
    return 0;
}